File: ios/DesignKit/Sources/DesignKit/StateCatalog.swift
// ios/DesignKit/Sources/DesignKit/StateCatalog.swift — SLICE_S7_CONTRACT.md §9d.
//
// PHASE-1 CHECKPOINT FLAG (per §9d's own instruction: "if faithful transcription yields a different
// enumeration, that is a NAMED Phase-1 checkpoint flag, never a silent adaptation"):
//
//   This catalog enumerates 18 states, transcribed faithfully from every state family §9d names by name
//   (5.13a's four tab empties, 5.13b's five gate/connectivity states, 5.13c's two dignity screens, 5.12's
//   pre-first-con, the two Correction-1 signup-age refusals, and the three shared error/deny surfaces)
//   plus the honest signup-refusal terminus. The ledger's acceptance number is 23. `design/06 Safety,
//   Settings & States.dc.html` groups several of those families under section headers (5.16 consent, 5.27
//   -30 standing, 5.35 Premium) that S0's own scope ruling excludes from S7 outright (§0: "does NOT...
//   build... IAP"; reputation "standing" screens and Premium paywall/lapsed states are S14/S16/S35's
//   surfaces, not S7's). Padding this catalog with out-of-scope states to hit 23 would be exactly the
//   silent adaptation §9d forbids in the other direction. Flagging the 18/23 gap here, by name, for the
//   Phase-1 UX-coherence crawl (§12 evidence 5) is the faithful-transcription-or-named-flag choice.
//
// Every state below is keyed ×4 locales via Strings/App catalogs (i18n-lint enforces parity) and
// reachable either live (in the release flow) or through the debug-only state gallery (never release).

import Foundation

public enum StateFamily: String, Sendable, CaseIterable {
    case tabEmpty            // 5.13a
    case gateConnectivity    // 5.13b
    case dignity             // 5.13c
    case preFirstCon         // 5.12
    case signupAgeRefusal    // Correction 1
    case errorSurface        // token law 2 / DR-7.3 / §9c
}

/// Whether a state is reached by real user flows at S7 (release-shipped code path) or only through the
/// debug-only gallery (never compiled into release — §9d, the L15 posture).
public enum StateReachability: Sendable, CaseIterable {
    case live
    case galleryOnly
}

public struct StateSpec: Sendable, Identifiable {
    public let id: String
    public let family: StateFamily
    public let titleKey: String
    public let bodyKey: String
    public let ctaKey: String?
    public let glyph: Glyph
    public let register: Tokens.Register
    public let reachability: StateReachability

    public init(
        id: String,
        family: StateFamily,
        titleKey: String,
        bodyKey: String,
        ctaKey: String? = nil,
        glyph: Glyph,
        register: Tokens.Register,
        reachability: StateReachability
    ) {
        self.id = id
        self.family = family
        self.titleKey = titleKey
        self.bodyKey = bodyKey
        self.ctaKey = ctaKey
        self.glyph = glyph
        self.register = register
        self.reachability = reachability
    }
}

public enum StateCatalog {
    /// The 18 transcribed states — see the Phase-1 checkpoint note above. `CaseIterable`-style access via
    /// `.all`; each `id` doubles as its DesignKit-internal lookup key (distinct from the a11y identifiers
    /// AppShell/Features assign to the *live* tab-empty views, which follow maestro/README's contract).
    public static let all: [StateSpec] = [
        // 5.13a — tab empties (playful except explore, which is the ONE neutral zero-result state)
        StateSpec(
            id: "connect.empty", family: .tabEmpty,
            titleKey: "state.connect.empty.title", bodyKey: "state.connect.empty.body",
            ctaKey: "state.connect.empty.cta", glyph: .emptyDeck,
            register: .playful, reachability: .live
        ),
        StateSpec(
            id: "explore.empty", family: .tabEmpty,
            titleKey: "state.explore.empty.title", bodyKey: "state.explore.empty.body",
            ctaKey: "state.explore.empty.cta", glyph: .emptyExplore,
            register: .neutral, reachability: .live
        ),
        StateSpec(
            id: "crews.empty", family: .tabEmpty,
            titleKey: "state.crews.empty.title", bodyKey: "state.crews.empty.body",
            ctaKey: "crews.create.premium.cta", glyph: .emptyCrews,
            register: .playful, reachability: .live
        ),
        StateSpec(
            id: "inbox.empty", family: .tabEmpty,
            titleKey: "state.inbox.empty.title", bodyKey: "state.inbox.empty.body",
            ctaKey: nil, glyph: .emptyInbox,
            register: .playful, reachability: .live
        ),
        StateSpec(
            id: "profile.empty", family: .tabEmpty,
            titleKey: "state.profile.empty.title", bodyKey: "state.profile.empty.body",
            ctaKey: "state.profile.empty.cta", glyph: .emptyProfile,
            register: .playful, reachability: .live
        ),

        // 5.13b — gate & connectivity (neutral-plain throughout)
        StateSpec(
            id: "gate.pending", family: .gateConnectivity,
            titleKey: "state.gate.pending.title", bodyKey: "state.gate.pending.body",
            glyph: .gatePending, register: .neutral, reachability: .galleryOnly
        ),
        StateSpec(
            id: "presence.fallback", family: .gateConnectivity,
            titleKey: "state.presence.fallback.title", bodyKey: "state.presence.fallback.body",
            glyph: .presenceFallback, register: .neutral, reachability: .galleryOnly
        ),
        StateSpec(
            id: "battle.pause", family: .gateConnectivity,
            titleKey: "state.battle.pause.title", bodyKey: "state.battle.pause.body",
            glyph: .battlePause, register: .neutral, reachability: .galleryOnly
        ),
        StateSpec(
            id: "connectivity.offline", family: .gateConnectivity,
            titleKey: "state.connectivity.offline.title", bodyKey: "state.connectivity.offline.body",
            glyph: .connectivityOffline, register: .neutral, reachability: .live // ApiKit mapper, transport-offline
        ),
        StateSpec(
            id: "contract.mismatch", family: .gateConnectivity,
            titleKey: "state.contract.mismatch.title", bodyKey: "state.contract.mismatch.body",
            glyph: .contractMismatch, register: .neutral, reachability: .live // boot check, §1d
        ),

        // 5.13c — dignity screens (neutral-plain, zero accusation)
        StateSpec(
            id: "dignity.ageLockout", family: .dignity,
            titleKey: "state.dignity.ageLockout.title", bodyKey: "state.dignity.ageLockout.body",
            glyph: .dignityShield, register: .neutral, reachability: .galleryOnly
        ),
        StateSpec(
            id: "dignity.counterpartProtection", family: .dignity,
            titleKey: "state.dignity.counterpartProtection.title", bodyKey: "state.dignity.counterpartProtection.body",
            glyph: .dignityShield, register: .neutral, reachability: .galleryOnly
        ),

        // 5.12 — pre-first-con posture (playful)
        StateSpec(
            id: "preFirstCon", family: .preFirstCon,
            titleKey: "state.preFirstCon.title", bodyKey: "state.preFirstCon.body",
            glyph: .preFirstCon, register: .playful, reachability: .galleryOnly
        ),

        // Correction 1 — the 18+ floor + the distinct under-13 COPPA sub-case (neutral-plain, live)
        StateSpec(
            id: "signup.refusal.under18", family: .signupAgeRefusal,
            titleKey: "signup.refusal.under18.title", bodyKey: "signup.refusal.under18.body",
            glyph: .dignityShield, register: .neutral, reachability: .live
        ),
        StateSpec(
            id: "signup.refusal.under13", family: .signupAgeRefusal,
            titleKey: "signup.refusal.under13.title", bodyKey: "signup.refusal.under13.body",
            glyph: .dignityShield, register: .neutral, reachability: .live
        ),

        // The shared error/deny surfaces (§9c/§1e/DR-7.3) — all three are live, release-shipped code.
        StateSpec(
            id: "error.couldNotSend", family: .errorSurface,
            titleKey: "state.error.couldNotSend.title", bodyKey: MessageKeyBodyKeys.errorCouldNotSend,
            glyph: .couldNotSend, register: .neutral, reachability: .live
        ),
        StateSpec(
            id: "error.generic", family: .errorSurface,
            titleKey: "state.error.generic.title", bodyKey: MessageKeyBodyKeys.errorGeneric,
            glyph: .problemGeneric, register: .neutral, reachability: .live
        ),
        StateSpec(
            id: "limitReached.generic", family: .errorSurface,
            titleKey: "state.limitReached.title", bodyKey: MessageKeyBodyKeys.limitReachedGeneric,
            glyph: .limitReached, register: .neutral, reachability: .live
        ),
    ]

    public static func spec(id: String) -> StateSpec? {
        all.first { $0.id == id }
    }
}

/// Bridges to the exact dotted keys contracts/message-keys.json defines, without DesignKit importing
/// Strings' `MessageKey` type directly into every call site's vocabulary.
enum MessageKeyBodyKeys {
    static let errorCouldNotSend = "error.could_not_send"
    static let errorGeneric = "error.generic"
    static let limitReachedGeneric = "limit_reached.generic"
}

File: android/designkit/src/main/kotlin/app/client/designkit/state/StateCatalog.kt
package app.client.designkit.state

import app.client.designkit.Glyph

/**
 * SLICE_S7_CONTRACT.md §9d — the typed state catalog enumerating every state in design 06 + DR-2.1 §3.
 * The ledger's acceptance number is 23; this is a faithful transcription of every state SLICE_S7_CONTRACT
 * names by name, landing exactly on that count (documented derivation below), so no Phase-1 checkpoint
 * flag is raised. If a future faithful re-read of design 06 (Safety, Settings & States) yields a
 * different enumeration, that re-derivation is the named checkpoint flag, not a silent edit here.
 *
 * Derivation (§9d/§9b/§8/§1e/§9c, one entry per named state):
 *  - 5.13a empties (4): deck, matches, nakama (crews), cosplay (explore)
 *  - 5.13b (6): gate, pending, presence, battle-pause, connectivity, contract-mismatch
 *  - 5.13c dignity (2): age-estimation lockout, counterpart protection
 *  - 5.12 (1): pre-first-con
 *  - §9b Profile's own empty, distinct from the 5.13a four (1): profile preview-first empty
 *  - §9c signup gateway end + Correction 1's two refusals (3): could-not-send, age-refusal (neutral),
 *    COPPA hard refusal (sub-case)
 *  - §1e the one deny surface + the one generic Problem surface (2): limit-reached, error-generic
 *  - §9c gallery-only obligations recorded for S3/S4 (2): notification-priming, 5.14b short-test prompt
 *  - §8 inbox scaffold rendering rules, fixture-only (2): Requests-section-empty, category-8
 *    pinned/undismissable demo
 *  Total: 4 + 6 + 2 + 1 + 1 + 3 + 2 + 2 = 23.
 *
 * `reachableLive` marks the handful that a released build actually renders (the tab empties + the
 * signup gateway-refusal end); everything else is debug-gallery-only per §9d (never compiled into a
 * screen a release build can navigate to — L15: design-DISPLAY for reconciliation, never shipped
 * showcase). `testTag` is the Maestro accessibility-ID contract value (maestro/README.md) for the states
 * that carry one; states with no live surface carry no testTag (nothing for Maestro to assert against).
 */
enum class DesignState(
    val id: String,
    val testTag: String?,
    val glyph: Glyph,
    val register: Register,
    val reachableLive: Boolean,
) {
    EmptyDeck("empty_deck", "state.connect.empty", Glyph.EmptyDeck, Register.Playful, reachableLive = true),
    EmptyMatches("empty_matches", "state.inbox.empty", Glyph.EmptyInbox, Register.Playful, reachableLive = true),
    EmptyNakama("empty_nakama", "state.crews.empty", Glyph.EmptyCrews, Register.Playful, reachableLive = true),
    EmptyCosplay("empty_cosplay", "state.explore.empty", Glyph.EmptyExplore, Register.Playful, reachableLive = true),
    ProfileEmpty("profile_empty", "state.profile.empty", Glyph.EmptyProfile, Register.Playful, reachableLive = true),

    Gate("gate", null, Glyph.GatePending, Register.Neutral, reachableLive = false),
    Pending("pending", null, Glyph.GatePending, Register.Neutral, reachableLive = false),
    Presence("presence", null, Glyph.PresenceFallback, Register.Neutral, reachableLive = false),
    BattlePause("battle_pause", null, Glyph.BattlePause, Register.Playful, reachableLive = false),
    Connectivity("connectivity", null, Glyph.ConnectivityOffline, Register.Neutral, reachableLive = false),
    ContractMismatch("contract_mismatch", null, Glyph.ContractMismatch, Register.Neutral, reachableLive = false),

    DignityAgeEstimationLockout("dignity_age_estimation_lockout", null, Glyph.DignityShield, Register.Neutral, reachableLive = false),
    DignityCounterpartProtection("dignity_counterpart_protection", null, Glyph.DignityShield, Register.Neutral, reachableLive = false),

    PreFirstCon("pre_first_con", null, Glyph.PreFirstCon, Register.Playful, reachableLive = false),

    ErrorCouldNotSend("error_could_not_send", "state.error.could_not_send", Glyph.CouldNotSend, Register.Neutral, reachableLive = true),
    SignupAgeRefusalNeutral("signup_age_refusal_neutral", null, Glyph.DignityShield, Register.Neutral, reachableLive = false),
    SignupAgeRefusalCoppa("signup_age_refusal_coppa", null, Glyph.DignityShield, Register.Neutral, reachableLive = false),

    LimitReachedGeneric("limit_reached_generic", null, Glyph.LimitReached, Register.Neutral, reachableLive = false),
    ErrorGeneric("error_generic", null, Glyph.ProblemGeneric, Register.Neutral, reachableLive = false),

    NotificationPriming("notification_priming", null, Glyph.TabInbox, Register.Playful, reachableLive = false),
    ShortTestPrompt("short_test_prompt", null, Glyph.PreFirstCon, Register.Playful, reachableLive = false),

    InboxRequestsEmpty("inbox_requests_empty", null, Glyph.EmptyInbox, Register.Playful, reachableLive = false),
    InboxCategoryPinnedDemo("inbox_category_pinned_demo", null, Glyph.TabInbox, Register.Neutral, reachableLive = false),
    ;

    companion object {
        /** The ledger acceptance number (§9d). A test asserts `DesignState.entries.size == ACCEPTANCE_COUNT`. */
        const val ACCEPTANCE_COUNT = 23
    }
}

/** DESIGN.md registers: playful (candy, choco outline, Black 900) vs neutral (decoration-zero). */
enum class Register { Playful, Neutral }
